### ISSUE TITLE
\tests\DCPS\RtpsDiscovery  no file called **TestMsgTypeSupportImpl.h**

### ISSUE THREAD

**Issue by Zhang8912** [objectcomputing/OpenDDS#3494]
Hi Eveyone,

I am a noob learner of OpenDDS and just want to know how Real Time Publisher and Subscriber find each other. 
I found an example \tests\DCPS\RtpsDiscovery, but here use a C++ header called **TestMsgTypeSupportImpl.h**, which is not saved in directory.
Do you know that where can I find this file? **TestMsgTypeSupportImpl.h**

thank you in advance

#include "TestMsgTypeSupportImpl.h"

**Comment by iguessthislldo** [comment]
Almost every reference in the source code in the form of `XTypeSupport.h`, `XTypeSupportImpl.h` or `XC.h` is a reference to a file generated by `tao_idl` and/or `opendds_idl` from an IDL file named `X.idl`. In this case it's referring to [`tests/DCPS/RtpsDiscovery/TestMsg.idl`](https://github.com/objectcomputing/OpenDDS/blob/master/tests/DCPS/RtpsDiscovery/TestMsg.idl). If you build the test, either separately or rebuilding OpenDDS with `configure --tests`, then `TestMsgTypeSupportImpl.h` will get generated.

As to how Publishers and Subscribers discover each other, I talked about about this as part of [an article from last year](https://objectcomputing.com/resources/publications/sett/october-2021-wireshark-and-opendds). From the source code perspective, the exact code for that is in  [`dds/DCPS/RTPS/Spdp.h`](https://github.com/objectcomputing/OpenDDS/blob/master/dds/DCPS/RTPS/Spdp.h) and [`dds/DCPS/RTPS/Spdp.cpp`](https://github.com/objectcomputing/OpenDDS/blob/master/dds/DCPS/RTPS/Spdp.cpp) (SPDP is the "Simple Participant Discovery Protocol").

The short version of this is that RTPS participants listen on a predetermined [multicast](https://en.wikipedia.org/wiki/Multicast) port, which can be heard from other devices on the network. Participants announce over these ports that they exist (or still exist) and these announcements eventually make their way to [this function](https://github.com/objectcomputing/OpenDDS/blob/469e395dc64e785009ce22d2b7760dbb39e7dec6/dds/DCPS/RTPS/Spdp.cpp#L688). If a participant hears about a new participant, then they both initiate a process where they tell each other about themselves and get their DataWriters and DataReaders talking to each other if they match up.
